Source/missiles.cpp

```diff
@@ -3724,39 +3688,83 @@ void ProcessRhino(Missile &missile)
 	PutMissile(missile);
 }
 
-void ProcessFireWallControl(Missile &missile)
+/*
+ * @brief Moves the control missile towards the right and grows walls.
+ * @param missile The control missile.
+ * @param sourceIdx The source index of the wall missile.
+ * @param type The missile ID of the wall missile.
+ * @param dmg The damage of the wall missile.
+ */
+static void ProcessWallControlLeft(Missile &missile, const int sourceIdx, const MissileID type, const int dmg)
+{
+	const Point leftPosition = { missile.var1, missile.var2 };
+	const Point target = leftPosition + static_cast<Direction>(missile.var3);
+
+	if (!missile.limitReached && GrowWall(sourceIdx, leftPosition, target, type, missile._mispllvl, dmg)) {
+		missile.var1 = target.x;
+		missile.var2 = target.y;
+	} else {
+		missile.limitReached = true;
+	}
+}
+
+/*
+ * @brief Moves the control missile towards the right and grows walls.
+ * @param missile The control missile.
+ * @param sourceIdx The source index of the wall missile.
+ * @param type The missile ID of the wall missile.
+ * @param dmg The damage of the wall missile.
+ * @param skipTile Should the tile be skipped.
+ */
+static void ProcessWallControlRight(Missile &missile, const int sourceIdx, const MissileID type, const int dmg, const bool skipTile = false)
+{
+	const Point rightPosition = { missile.var5, missile.var6 };
+	const Point target = rightPosition + static_cast<Direction>(missile.var4);
+
+	if (missile.var7 == 0 && GrowWall(sourceIdx, rightPosition, target, type, missile._mispllvl, dmg, skipTile)) {
+		missile.var5 = target.x;
+		missile.var6 = target.y;
+	} else {
+		missile.var7 = 1;
+	}
+}
+
+void ProcessWallControl(Missile &missile)
 {
 	missile._mirange--;
 	if (missile._mirange == 0) {
 		missile._miDelFlag = true;
 		return;
 	}
 
-	int id = missile._misource;
-
-	{
-		Point position = { missile.var1, missile.var2 };
-		Point target = position + static_cast<Direction>(missile.var3);
+	MissileID type;
 
-		if (!missile.limitReached && GrowWall(id, position, target, MissileID::FireWall, missile._mispllvl, 0)) {
-			missile.var1 = target.x;
-			missile.var2 = target.y;
-		} else {
-			missile.limitReached = true;
-		}
+	switch (missile._mitype) {
+	case MissileID::FireWallControl:
+		type = MissileID::FireWall;
+		break;
+	case MissileID::LightningWallControl:
+		type = MissileID::LightningWall;
+		break;
+	default:
+		type = MissileID::Null;
+		break;
 	}
 
-	{
-		Point position = { missile.var5, missile.var6 };
-		Point target = position + static_cast<Direction>(missile.var4);
+	const int sourceIdx = missile._misource;
+	int lvl = 0;
+	int dmg = 0;
 
-		if (missile.var7 == 0 && GrowWall(id, position, target, MissileID::FireWall, missile._mispllvl, 0)) {
-			missile.var5 = target.x;
-			missile.var6 = target.y;
-		} else {
-			missile.var7 = 1;
-		}
+	if (type == MissileID::LightningWall) {
+		lvl = !missile.IsTrap() ? Players[sourceIdx].getCharacterLevel() : 0;
+		dmg = 16 * (GenerateRndSum(10, 2) + lvl + 2);
 	}

```

Comment:
Can we include this logic in the new switch above?

Reply:
Yes